src/commands/git-metadata/gitdb.ts

```diff
@@ -0,0 +1,281 @@
+import FormData from 'form-data'
+import * as simpleGit from 'simple-git'
+
+import {RequestBuilder} from '../../helpers/interfaces'
+import {retryRequest} from '../../helpers/retry'
+
+import {gitRemote} from './git'
+import {Logger} from './utils'
+
+const API_TIMEOUT = 15000
+
+export const uploadToGitDB = async (
+  log: Logger,
+  request: RequestBuilder,
+  git: simpleGit.SimpleGit,
+  dryRun: boolean
+) => {
+  let repoURL
+  try {
+    repoURL = await getRepoURL(git)
+    log.debug(`Syncing repository ${repoURL}`)
+  } catch (err) {
+    log.warn(`Failed getting repository URL: ${err}`)
+    throw err
+  }
+
+  let latestCommits
+  try {
+    latestCommits = await getLatestLocalCommits(git)
+    if (latestCommits.length === 0) {
+      log.debug('No local commits found.')
+
+      return
+    }
+    log.debug(`${latestCommits.length} commits found, asking GitDB which ones are missing.`)
+  } catch (err) {
+    log.warn(`Failed getting local commits: ${err}`)
+    throw err
+  }
+
+  let commitsToExclude
+  try {
+    commitsToExclude = await getKnownCommits(log, request, repoURL, latestCommits)
+    log.debug(`${commitsToExclude.length} commits already in GitDB.`)
+  } catch (err) {
+    log.warn(`Failed getting commits to exclude: ${err}`)
+    throw err
+  }
+
+  let commitsToUpload
+  try {
+    commitsToUpload = await getCommitsToUpload(git, commitsToExclude)
+    log.debug(`${commitsToUpload.length} commits to upload.`)
+  } catch (err) {
+    log.warn(`Failed getting commits to upload: ${err}`)
+    throw err
+  }
+
+  let packfiles
+  try {
+    packfiles = generatePackFilesForCommits(log, commitsToUpload)
+    log.debug(`${packfiles.length} packfiles generated.`)
+  } catch (err) {
+    log.warn(`Failed generating packfiles: ${err}`)
+    throw err
+  }
+
+  if (dryRun) {
+    log.debug(`Dry-run enabled, not uploading anything.`)
+
+    return
+  }
+  log.debug(`Uploading packfiles...`)
+  try {
+    await uploadPackfiles(log, request, repoURL, latestCommits[0], packfiles)
+    log.debug(`Successfully uploaded packfiles.`)
+  } catch (err) {
+    log.warn(`Failed to upload packfiles: ${err}`)
+    throw err
+  }
+}
+
+const getRepoURL = gitRemote
+
+const getLatestLocalCommits = async (git: simpleGit.SimpleGit) => {
+  const logResult = await git.log(['-n 1000', '--since="1 month ago"'])
+
+  return logResult.all.map((c) => c.hash)
+}
+
+// getKnownCommits asks the backend which of the given commits are already known
+const getKnownCommits = async (log: Logger, request: RequestBuilder, repoURL: string, latestCommits: string[]) => {
+  interface SearchCommitResponse {
+    data: Commit[]
+  }
+
+  interface Commit {
+    type: string
+    id: string
+  }
+
+  const localCommitData = JSON.stringify({
+    meta: {
+      repository_url: repoURL,
+    },
+    data: latestCommits.map((commit) => ({
+      id: commit,
+      type: 'commit',
+    })),
+  })
+  const response = await runRequest(log, 'search_commits', () =>
```

Comment:
Should we check the status code of the response before trying to read the body?

Reply:
This is done by axios, which will throw an exception if the response is not in the 2xx range ([see here the default behavior](https://github.com/axios/axios/blob/v0.21.4/README.md?plain=1#L410) which is what is configured since we don't override this option)